Fix index not showing in fileserver

Zuyoutoki opened the pull request:
At the root of the repo, run the following:
```sh
echo "index file content" > src/index.html
zig build example -Dexample=router
```
In a new terminal:
```sh
curl http://127.0.0.1:8080/files/
```
We get `Resource not found` instead of `index file content`.

The issue is also there with subdirectories:
```sh
mkdir src/subdir
echo "index in a subdirectory" > src/subdir/index.html
zig build example -Dexample=router
```
```sh
curl http://127.0.0.1:8080/files/subdir/
```

## Fix
This PR fix the issue by serving the index file when the query ask for a directory.

Luukdegram commented on src/fs.zig:
```diff
@@ -67,7 +67,23 @@ pub fn serve(ctx: void, response: *Response, request: Request) ServeError!void {
+            else => |e| return e,
+        };
+        if (stat.kind == .Directory) {
+            break :blk try std.mem.concat(alloc, u8, &.{ path, "/", index });
```
This leaks memory as the result is never freed. The best solution is to have a buffer with the size `std.os.MAX_PATH_BYTES` and use that buffer to backup the concatenation using memcpy. (similarly to line 52).

Luukdegram commented on src/fs.zig:
```diff
@@ -67,7 +67,23 @@ pub fn serve(ctx: void, response: *Response, request: Request) ServeError!void {
+            break :blk try std.mem.concat(alloc, u8, &.{ path, "/", index });
+        }
+        // otherwise, we should serve the file at path
+        else {
+            break :blk path;
+        }
```
This can be simplified to `break :blk path;`, no need for the else branch.

Luukdegram reviewed (changes requested):
Thank you for the contribution. Just some small changes I'd like to see made. It would also be really appreciated if a test case could be added, though not mandatory.

Zuyoutoki commented:
You're welcome! I've submitted the changes requested. I included a test case inspired by the one in `server.zig`, but I think some sort of race condition makes it send only the status and headers with no body every once in a while. I don't have much experience debugging race conditions, but I'll try to figure it out in the next few days. If you have some insight about it, let me know, any help is welcome :smile:

Luukdegram commented on src/fs.zig:
```diff
@@ -159,3 +157,75 @@ pub fn serveFile(
+
+    // setup files to serve
+    const cwd = std.fs.cwd();
+    cwd.makeDir("tmp-dir-for-tests") catch |err| switch (err) {
```
You can actually use `std.testing.tmpDir` for this. This will make sure we do not dirty the working directory of the user in the event something goes wrong and the delete action didn't work.

Luukdegram reviewed (changes requested):
The server in the test doesn't enable the fileserver, so it doesn't contain the logic to default to `index.html` when the path is empty.

Zuyoutoki commented:
I moved the fileserver init inside the server_thread (was that what you meant by your last comment?) and changed the files setup to use `std.testing.tmpDir`. Unfortunately, this didn't solve the issue.

After some more testing, the issue seems to boil down to this : `serveFile` flush the status+headers, then `sendfile` writes the file content to the stream, so we have 2 writes. From my understanding, `stream.reader().read` will wait for at least a byte before returning and will read all it can, so depending on when we call it, `sendfile` might not have been called yet, therefore why the test fails to read the file content. A second read should give us the missing part.

To make sure this was what caused the issue, I temporarily inserted `std.time.nanosleep(1_000_000);` before `sendfile` and was able to get the error every time. When I inserted the sleep before `stream.reader().read`, I was able to run the tests successfully every time.

I could think of two ways to fix the issue. The first one would be to let it sleep, but it seems hacky so I went with the other one. The second one is to make the test client read the `Content-Length` header and call `stream.reader().read` again if the received length is not right. I also included a retry count in the while loop to prevent deadlocks. 

Hopefully, the test should work fine now :smile:

Luukdegram commented:
Great changes! Thank you very much for your contribution.
